### PR TITLE
feat: enable users to listen for token refresh event

### PR DESCRIPTION
**Pull request by harrysolovay**
_Issue:_ #6352 

_Description of changes:_

Enables users to do the following.

```ts
Hub.listen('auth', (data) => {
  switch (data.payload.event) {
    case 'token_refresh': {
      logger.log("New token has been process");
    }
  }
});
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -84,6 +84,19 @@ const dispatchAuthEvent = (event: string, data: any, message: string) => {
...
 };
 
+const eventEmitters = {
+	onTokenRefresh() {
+		dispatchAuthEvent('token_refresh', undefined, `New token retrieved`);
+	},
+	onTokenRefreshFailure(error: Error) {
+		dispatchAuthEvent(
+			'token_refresh_failure',
+			error,
```
Comment (sammartinez) [inline packages/auth/src/Auth.ts:94]: Can you provide an example of this `error`?

Comment (harrysolovay) [reply packages/auth/src/Auth.ts:94]: Could be any of [these](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/CommonErrors.html), for instance.

File: packages/auth/src/Auth.ts [hunk 2]
```diff
@@ -84,6 +84,19 @@ const dispatchAuthEvent = (event: string, data: any, message: string) => {
 	Hub.dispatch('auth', { event, data, message }, 'Auth', AMPLIFY_SYMBOL);
 };
 
+const eventEmitters = {
+	onTokenRefresh() {
+		dispatchAuthEvent('token_refresh', undefined, `New token retrieved`);
```
Comment (elorzafe) [inline packages/auth/src/Auth.ts:89]: What will happen when destructuring `undefined`?

Comment (elorzafe) [reply packages/auth/src/Auth.ts:89]: nvm

**Pull request closed without merging**